Sources/MongoSwift/SDAM.swift

```diff
@@ -24,11 +33,45 @@ public struct ServerAddress: Equatable {
 
     /// Initializes a ServerAddress, using the default localhost:27017 if a host/port is not provided.
     internal init(_ hostAndPort: String = "localhost:27017") throws {
-        let parts = hostAndPort.split(separator: ":")
-        self.host = String(parts[0])
-        guard let port = UInt16(parts[1]) else {
-            throw MongoError.InternalError(message: "couldn't parse address from \(hostAndPort)")
+        guard !hostAndPort.contains("?") else {
+            throw MongoError.InvalidArgumentError(message: "\(hostAndPort) contains invalid characters")
         }
+
+        var host: String
+        var port: UInt16
+
+        // Check if host is an IPv6 literal.
+        let ipLiteralRegex = try NSRegularExpression(pattern: #"^\[(.*)\](?::([0-9]+))?$"#)
+        if hostAndPort.first == "[" {
+            guard let match = ipLiteralRegex.firstMatch(
+                in: hostAndPort,
+                range: NSRange(location: 0, length: hostAndPort.utf16.count)
```

Comment:
Strings in Swift are pretty complicated, since they can represent a lot more than just ASCII characters and can be viewed in multiple different encodings. In this case, we're counting how long the UTF-16 encoded version of the string is, but this is most of the time different than how many characters there are in the string. The count of the UTF-8 representation often does reflect the number of characters, but sometimes that is also off if the string contains any multibyte characters. To handle this situation, Swift Strings use abstract character indexes which can be used independent of any encoding of the underlying data. We can use such indexes to define a range that properly spans the entirety of the string, regardless of any weird characters in it.

e.g.
```swift
NSRange(data.startIndex..<data.endIndex, in: data)
```
For more information on this, the documentation of the `String` type has a good overview: https://developer.apple.com/documentation/swift/string

Reply:
This still needs to be updated from the UTF-16 count to use the indexes.

Reply:
Just to ping on this again, it looks like this still needs to be updated.